### DISCUSSION TITLE
Dagster not processing ticks while another is in progress, doesn't catch up on missed ticks.

### DISCUSSION THREAD

**Discussion by NiallRees** [dagster-io/dagster#20078] [Q&A] [answered]
(Dagster version 1.6.6, Helm on GKE)

Hi there, we're using an hourly schedule definition with logic to return the required tenants that need to be synced each hour. The majority of the tenants sync at 3AM UTC. The issue we're encountering is that this 3AM tick is blocking downstream ticks from running. This is happening both while the 3AM tick is still in progress, but also once it has completed. Dagster isn't catching up on the ticks that were missed.

While the 3AM tick is in progress (note the missing tick entries between 3AM and 9AM)
<img width="1728" alt="image" src="https://github.com/dagster-io/dagster/assets/23722609/264e15c8-15a3-418e-b707-e8d6501643b7">

Once the 3AM tick has completed (note that ticks pick up at 9AM again after the tick finished, but there are no entries for 4AM, 5AM, 6AM, 7AM or 8AM)
<img width="1387" alt="image" src="https://github.com/dagster-io/dagster/assets/23722609/2c9ed80c-3e8b-4ab3-86e9-fcf87ebb9bc8">

Is there a setting to tell Dagster to catch up and not skip those ticks, or is this a bug? Thanks.

**Comment by gibsondan** [comment]
Hi @NiallRees - Dagster is designed around schedule ticks being relatively lightweight computations that figure out which runs to launch, with the heavy work happening in those launched runs. By default there's a 60 second timeout on schedule executions - it's possible to increase that timeout using environment variables (I imagine you had to do that in order for a 7 hour tick to be possible?) but if you're expecting a schedule tick to run for several hours we would recommend launching a run that does that longer work instead.

Re: not catching up - the scheduler will try to catch up if the job is partitioned (e.g. using build_schedule_from_partitioned_job: https://docs.dagster.io/concepts/partitions-schedules-sensors/schedules#time-partitioned-jobs-and-assets) - but it doesn't try to catch up if it's a regular cron schedule operating on a job with no underlying partitions.

**Reply by NiallRees** [reply]
Hey @gibsondan thanks a lot for the response. The schedule tick actually is a very lightweight computation, which is why I'm confused here. I think Dagster is showing 'evaluating' to mean that the schedule's runs are still in progress, rather than it being blocked on the schedule definition? The schedule tick is simply taking the list of partitions, and checking which should run with a database query.

Our job is partitioned, so not sure why it's not catching up.
```
replicator = define_asset_job(
    "replicator",
    selection=asset_selection,
    partitions_def=account_id_partitions,
    config={"execution": {"config": {"multiprocess": {"max_concurrent": 6}}}},
)
```

**Reply by gibsondan** [reply]
Got it - something strange is going on then. I can say with near certainty that "evaluating" doesn't mean that the runs are in progress, but rather that the daemon is still working through that particular schedule tick to figure out which run sneed to be launched.

Running py-spy on your daemon pod might give more clues for what is going on during that time and why it's taking so long: https://github.com/dagster-io/dagster/discussions/14771 - it's possible that some part of the system (the daemon or possibly your postgres DB) are overloaded and dramatically slowing things down.

Logs from your daemon pod during this time might also help explain what it thinks it's doing during those 7 hours.

How are you creating your schedule? The "Partition Set: None" in your screenshot indicates that it doesn't think the schedule has an underlying set of partitions, which might help explain why it isn't trying to catch up.

**Reply by NiallRees** [reply]
I've attached some logs in this other discussion I started - it felt like a separate issue but probably related: https://github.com/dagster-io/dagster/discussions/20082


Ok thanks I'm going to look into why that Partition set is not active for the schedule and see if that sorts out this specific discussion.

**Reply by gibsondan** [reply]
@NiallRees i'm actually seeing some indication that this might not be behaving as expected for certain schedules defined from partitioned jobs - so it's possible this is our bug, we'll dig into that and see what's going on there, thanks for the report.

**Reply by gibsondan** [reply]
Despite that while we sort that out, I think the root problem here is the tick taking 7 hours - as long as the schedule function executes in under an hour, which is more than an order of magnitude more than its typically allowed to run before timing out, you shouldn't need to worry about catchup behavior

**Reply by NiallRees** [reply]
Ah ok that makes sense. Yeah see the other discussion I linked. It’s taking 3 minutes to start each partitioned run from the schedule, so consequently it’s taking 150 * 3 minutes to finish launching the schedule

**Reply by NiallRees** [reply]
The schedule function is very fast as I said, but dagster is taking a long time to work through the runs it yields

**Reply by NiallRees** [reply]
Hey @gibsondan - we just had a go at deleting a ton of historic data in the Dagster db to see if it solves this. It doesn't unfortunately. If I test the schedule in the UI, it returns in less than a second with 150 run requests, so it's definitely not the schedule definition that's an issue. I am getting a ton of query timeouts in the UI though while the schedule is running.

**Reply by gibsondan** [reply]
Some things to consider/try:

- Running `py-spy record` on the daemon during a period of time when it's slow is the next step I'd recommend to troubleshoot this to identify the place where it's spending all the unexpected time - there is a guide to doing that here, you'll need to ensure that your daemon pod is running using the right security context to be able to run py-spy on it: https://github.com/dagster-io/dagster/discussions/14771 

- If you're also getting query timeouts it most likely indicates that your database is overloaded - understanding which query is running slow from py-spy will help understand the problem. 

- make sure that your database is up to date schema/migration-wise: if you run this query on your db:
```
SELECT * FROM alembic_version;
```
the result should be "46b412388816" if you're up to date. If it's not, there are instructions for migrating your DB on kubernetes here: https://docs.dagster.io/deployment/guides/kubernetes/how-to-migrate-your-instance#migrating-a-dagster-instance-while-upgrading

- Check the performance stats on your database and see if it's maxed out on CPU or memory - sometimes just bumping up the instance size is enough to resolve these kinds of problems.

- You mentioned cleaning up historical data in the DB - the main data that the scheduler queries in the DB during submitting runs is runs and run tags, so deleting old runs could potentially help: https://github.com/dagster-io/dagster/discussions/12047

- You might also find this numSubmitWorkers field helpful to submit the multiple runs in parallel, although it's unlikely to help fix whatever is making the database run extremely slowly: https://github.com/dagster-io/dagster/blob/master/helm/dagster/values.yaml#L1142

**Reply by NiallRees** [reply]
Thank you @gibsondan this is very helpful - appreciate the detail. We've just completely recreated the database as a next step, to rule out any bad index issues. If that hasn't sorted it I'll go for the py-spy route.

**Reply by NiallRees** [reply, marked as answer]
I recreated the DB, which didn't do anything. Doubling the DB CPU has solved it. It still makes me uncomfortable that I don't know exactly why though 😅. Today it took 18 minutes to launch all 150 runs, vs the screenshot initially where it was at 6 hours! Really weird. Thanks for your help @gibsondan - if it gets weird again and to a point of prohibitive db expense I'll definitely dig out py-spy.

<img width="1134" alt="image" src="https://github.com/dagster-io/dagster/assets/23722609/1b3ce86d-a9b3-4e62-8e27-26594050fd7c">
